File: sdks/python/src/autonoma/dialect.py
"""Database dialect abstraction — generates dialect-specific SQL strings."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Protocol

from .generated.sql_queries import (
    POSTGRES_TABLES,
    POSTGRES_COLUMNS,
    POSTGRES_PRIMARY_KEYS,
    POSTGRES_FOREIGN_KEYS,
    POSTGRES_ENUMS,
    MYSQL_TABLES,
    MYSQL_COLUMNS,
    MYSQL_PRIMARY_KEYS,
    MYSQL_FOREIGN_KEYS,
    MYSQL_ENUMS,
)


def _replace_schema(template: str, schema: str) -> str:
    return template.replace("{{schema}}", schema)


class Dialect(Protocol):
    name: str
    supports_returning: bool

    def param(self, index: int) -> str: ...
    def quote_id(self, name: str) -> str: ...
    def tables_sql(self, schema: str) -> str: ...
    def columns_sql(self, schema: str) -> str: ...
    def primary_keys_sql(self, schema: str) -> str: ...
    def foreign_keys_sql(self, schema: str) -> str: ...
    def enums_sql(self, schema: str) -> str: ...


@dataclass
class PostgresDialect:
    name: str = "postgres"
    supports_returning: bool = True

    def param(self, index: int) -> str:
        return f"${index}"

    def quote_id(self, name: str) -> str:
        return f'"{name}"'

    def tables_sql(self, schema: str) -> str:
        return _replace_schema(POSTGRES_TABLES, schema)

    def columns_sql(self, schema: str) -> str:
        return _replace_schema(POSTGRES_COLUMNS, schema)

    def primary_keys_sql(self, schema: str) -> str:
        return _replace_schema(POSTGRES_PRIMARY_KEYS, schema)

    def foreign_keys_sql(self, schema: str) -> str:
        return _replace_schema(POSTGRES_FOREIGN_KEYS, schema)

    def enums_sql(self, _schema: str) -> str:
        return POSTGRES_ENUMS


@dataclass
class MySQLDialect:
    name: str = "mysql"
    supports_returning: bool = False

    def param(self, _index: int) -> str:
        return "?"

    def quote_id(self, name: str) -> str:
        return f"`{name}`"

    def tables_sql(self, schema: str) -> str:
        return _replace_schema(MYSQL_TABLES, schema)

    def columns_sql(self, schema: str) -> str:
        return _replace_schema(MYSQL_COLUMNS, schema)

    def primary_keys_sql(self, schema: str) -> str:
        return _replace_schema(MYSQL_PRIMARY_KEYS, schema)

    def foreign_keys_sql(self, schema: str) -> str:
        return _replace_schema(MYSQL_FOREIGN_KEYS, schema)

    def enums_sql(self, _schema: str) -> str:
        return MYSQL_ENUMS


def get_dialect(name: str = "postgres") -> PostgresDialect | MySQLDialect:
    if name == "postgres":
        return PostgresDialect()
    elif name == "mysql":
        return MySQLDialect()
    else:
        raise ValueError(f'Dialect "{name}" is not yet supported. Currently only "postgres" and "mysql" are available.')

File: sdks/ruby/lib/autonoma/dialect.rb
# frozen_string_literal: true

require_relative "generated/sql_queries"

module Autonoma
  module Dialect
    def self.get_dialect(name = "postgres")
      case name
      when "postgres"
        PostgresDialect.new
      when "mysql"
        MySQLDialect.new
      else
        raise ArgumentError, "Dialect \"#{name}\" is not yet supported. Currently only \"postgres\" and \"mysql\" are available."
      end
    end
  end

  class PostgresDialect
    def name
      "postgres"
    end

    def supports_returning
      true
    end

    def param(index)
      "$#{index}"
    end

    def quote_id(name)
      "\"#{name}\""
    end

    def tables_sql(schema)
      replace_schema(SqlQueries::POSTGRES_TABLES, schema)
    end

    def columns_sql(schema)
      replace_schema(SqlQueries::POSTGRES_COLUMNS, schema)
    end

    def primary_keys_sql(schema)
      replace_schema(SqlQueries::POSTGRES_PRIMARY_KEYS, schema)
    end

    def foreign_keys_sql(schema)
      replace_schema(SqlQueries::POSTGRES_FOREIGN_KEYS, schema)
    end

    def enums_sql(_schema)
      SqlQueries::POSTGRES_ENUMS
    end

    private

    def replace_schema(template, schema)
      template.gsub("{{schema}}", schema)
    end
  end

  class MySQLDialect
    def name
      "mysql"
    end

    def supports_returning
      false
    end

    def param(_index)
      "?"
    end

    def quote_id(name)
      "`#{name}`"
    end

    def tables_sql(schema)
      replace_schema(SqlQueries::MYSQL_TABLES, schema)
    end

    def columns_sql(schema)
      replace_schema(SqlQueries::MYSQL_COLUMNS, schema)
    end

    def primary_keys_sql(schema)
      replace_schema(SqlQueries::MYSQL_PRIMARY_KEYS, schema)
    end

    def foreign_keys_sql(schema)
      replace_schema(SqlQueries::MYSQL_FOREIGN_KEYS, schema)
    end

    def enums_sql(_schema)
      SqlQueries::MYSQL_ENUMS
    end

    private

    def replace_schema(template, schema)
      template.gsub("{{schema}}", schema)
    end
  end
end
